test/test_controller.py

```diff
@@ -0,0 +1,177 @@
+import pytest
+import json
+
+from larpix import Controller, Configuration_v2
+from larpix.io import FakeIO
+
+@pytest.fixture
+def network_config(tmpdir):
+    '''
+    Defines a network of:
+        MISO_US:
+            3       13
+            ^       ^
+            2   >   12
+            ^
+        MISO_DS:
+            3       13
+            v       v
+            2   <   12
+            v
+
+    '''
+    filename = str(tmpdir.join('test_network_conf.json'))
+    config_dict = {
+        "_config_type": "controller",
+        "name": "test",
+        "type": "network",
+        "network": {
+            "1": { "1": { "chips": [
+                        {"chip_id": 2,
+                        "miso_us": [3, None, None, 12],
+                        "root": True,
+                        "miso_ds": ["ext", None, None, None]},
+                        {"chip_id": 3},
+                        {"chip_id": 12,
+                        "miso_us": [ 13, None, None, None]},
+                        {"chip_id": 13}
+                    ]
+                }
+            },
+            "miso_us_uart_map": [0,1,2,3],
+            "miso_ds_uart_map": [2,3,0,1],
+            "mosi_uart_map": [0,1,2,3]
+        }
+    }
+    with open(filename,'w') as of:
+        json.dump(config_dict, of)
+    return filename
+
+@pytest.fixture
+def network_controller(network_config):
+    c = Controller()
+    c.load(network_config)
+    c.io = FakeIO()
+    return c
+
+def test_controller_network(network_controller):
+    c = network_controller
+    networks = ('miso_us','miso_ds','mosi')
+    for chip_key in c.chips:
+        for network in networks:
+            assert chip_key.chip_id in c.network[chip_key.io_group][chip_key.io_channel][network].nodes()
+    us_links = [(2,3),(2,12),(12,13)]
+    ds_links = [(3,2),(12,2),(13,12),(2,'ext')]
+    mosi_links = [link[::-1] for link in us_links] + [link[::-1] for link in ds_links]
+    assert all([link in c.network[chip_key.io_group][chip_key.io_channel]['miso_us'].edges() for link in us_links])
```

Comment:
It looks like this is comparing ``...edges()`` and ``us_links`` to see if they have the same elements. The cleanest way I've seen to do this is ``assert set(...edges()) == set(us_links)``.